File: cw3q2/scripts/iiwa14Kine_node.py
#!/usr/bin/env python

import rospkg
from math import pi
import rosbag
import rospy
import tf2_kdl
import tf2_ros

from cw3q2.iiwa14Kine import iiwa14_kinematic

if __name__ == '__main__':
    rospy.init_node('cw3q1_node')
    iiwa14 = iiwa14_kinematic()
    
    p0 = [0, 0, 0, 0, 0, 0, 0]
    p1 = [0.768506, 0.121759, -0.939627, 1.99111, -0.200346, -1.35671, -0.790215]
    v1 = [-0.033846, 0.0564997, 0.295151, 0.271778, 0.259246, 0.0124612, -0.0292725]

    p2 = [-1.58425, -0.272908, 0.796272, 2.37887, 0.0842769, -0.0436095, 1.05386]
    v2 = [-0.327998, -0.166912, 0.147049, -0.339939, -0.0167218, -0.272989, 0.0179721]

    p3 = [-2.13286, 0.266082, 0.942719, 2.37668, -2.04084, -0.598984, -0.832711]
    v3 = [0.203643, 0.130777, 0.0819627, 0.181411, 0.033077, -0.00593176, 0.0985884]

    #JOINT1
    T = iiwa14.forward_kine(p1, 6)
    T_cm = iiwa14.forward_kine_cm(p1, 6)

    jacobian_cm = iiwa14.get_jacobian_cm(p1,6)
    IK_ite = iiwa14.inverse_kine_ite(T, p0)
    T_ite = iiwa14.forward_kine(IK_ite, 6)
    B = iiwa14.getB(p1)
    C = iiwa14.getC(p1, v1)
    G = iiwa14.getG(p1)
    print("///// JOINT 1 /////")
    print("Position: ")
    print(p1)
    print("Velocity: ")
    print(v1)
    print("T_cm is: ")
    print(T_cm)

    print("jacobian_cm: ")
    print(jacobian_cm)
    print("The iterative form IK is: ")
    print(IK_ite)
    print("B is: ")
    print(B)
    print("C is: ")
    print(C)
    print("G is: ")
    print(G)
    print("\n")


    

File: cw3q2/src/iiwa14Kine_node.cpp
#include "ros/ros.h"
#include <tf2_eigen/tf2_eigen.h>
#include <Eigen/Dense>
#include <sensor_msgs/JointState.h>
#include "cw3q2/iiwa14Kine.h"
#include "random"
#include "chrono"

VectorXd generate_position()
{
    VectorXd random_joint;

    random_joint.resize(7);

    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::default_random_engine generator(seed);

    std::normal_distribution<double> dist(0.0, 5*M_PI/9);

    for (int i = 0; i < 7; i++)
        random_joint(i) = dist(generator);

    return random_joint;
}

VectorXd generate_velocity()
{
    VectorXd random_joint;

    random_joint.resize(7);

    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::default_random_engine generator(seed);

    std::normal_distribution<double> dist(0.0, M_PI/15);

    for (int i = 0; i < 7; i++)
        random_joint(i) = dist(generator);

    return random_joint;
}


void iiwa14_kinematics() {

    Eigen::Matrix4d A;
    //double curr_joint[5], joint1_values[5], joint2_values[5], joint3_values[5], joint4_values[5], joint5_values[5];

    Matrix4d T;
    Eigen::MatrixXd T_cm, T_ite, jacobian, jacobian_cm, IK_closed, B, C, G;
    Eigen::VectorXd IK_ite, p0(7), p1(7), v1(7), p2(7), v2(7), p3(7), v3(7);
	
	iiwa14_kinematic object;
	object.init();
	
	p0 << 0, 0, 0, 0, 0, 0, 0;
	//first test point
	p1 << 0.768506, 0.121759, -0.939627, 1.99111, -0.200346, -1.35671, -0.790215;
	v1 << -0.033846, 0.0564997, 0.295151, 0.271778, 0.259246, 0.0124612, -0.0292725;
	//second test point
	p2 << -1.58425, -0.272908, 0.796272, 2.37887, 0.0842769, -0.0436095, 1.05386;
	v2 << -0.327998, -0.166912, 0.147049, -0.339939, -0.0167218, -0.272989, 0.0179721;
	//third test point
	p3 << -2.13286, 0.266082, 0.942719, 2.37668, -2.04084, -0.598984, -0.832711;
	v3 << 0.203643, 0.130777, 0.0819627, 0.181411, 0.033077, -0.00593176, 0.0985884;


	T_cm = object.forward_kine_cm(p1,7);
	jacobian = object.get_jacobian(p1);
	jacobian_cm = object.get_jacobian_cm(p1,7);
	B = object.getB(p1);
	C = object.getC(p1, v1);
	G = object.getG(p1);

	std::cout << "///// JOINT 1 /////" << std::endl;
	std::cout << "Position: \n" << "[" << p1[0] << ", " << p1[1] << ", " << p1[2] << ", " << p1[3] << ", " << p1[4] << ", " << p1[5] << ", " << p1[6] << "]" << "\n" << std::endl;
	std::cout << "Velocity: \n" << "[" << v1[0] << ", " << v1[1] << ", " << v1[2] << ", " << v1[3] << ", " << v1[4] << ", " << v1[5] << ", " << v1[6] << "]" << "\n" << std::endl;
	std::cout << "T_cm is: \n"<< T_cm << "\n" << std::endl;
	std::cout << "The Jacobian is: \n"<< jacobian << "\n" << std::endl;
	std::cout << "The Jacobian at the cm is: \n"<< jacobian_cm << "\n" << std::endl;
	std::cout << "The iterative form IK is: \n"<< IK_ite << std::endl;
	std::cout << "B: \n"<< B << "\n" << std::endl;
	std::cout << "C: \n"<< C << "\n" << std::endl;
	std::cout << "G: \n"<< G << "\n" << std::endl;
	std::cout << "\n" << std::endl;

}

int main(int argc, char **argv)
{	
    ros::init(argc, argv, "cw3q2_node");
    //ros::NodeHandle nh;

    iiwa14_kinematics();
    ros::spin();
}
